### PR TITLE
BUG: interpolate: sanity check x and y in make_interp_spline(x, y, k=0) and k=1

### PR DESCRIPTION
**Pull request by ev-br**
#### What does this implement/fix?

`make_interp_spline` short-circuits for k=0 and k=1 (piecewise constant and linear interpolation). And it does it *before* sanity checking the input arrays. This allows passing in e.g. x and y arrays of inconsistent lengths---which k>=2 modes do not allow.

Thus move the sanity checks to be earlier and add tests to assert consistent behavior of .interpolators of various degrees.

#### Additional information

The fact that this was not caught for years hints at how often this is used :-). However this was masking a separate issue in gh-16508 (which proves to be surprisingly effective in smoking out obscure problems :-)), so it makes sense to fix it separately.

### CODE REVIEW

**Comment by ev-br** [comment]
@j-bowhay would you be able to review this please?

**Review by j-bowhay — APPROVED** [review]
LGTM, failures are unrelated and the change is straight forward

File: scipy/interpolate/tests/test_bsplines.py [hunk 1]
```diff
@@ -942,6 +942,29 @@ def test_linear(self):
         b = make_interp_spline(self.xx, self.yy, k=1, axis=-1)
         assert_allclose(b(self.xx), self.yy, atol=1e-14, rtol=1e-14)
 
+    @pytest.mark.parametrize('k', [0, 1, 2, 3])
+    def test_incompatible_x_y(self, k):
+        x = [0, 1, 2, 3, 4, 5]
+        y = [0, 1, 2, 3, 4, 5, 6, 7]
+        with assert_raises(ValueError):
```
Comment (tupui) [inline scipy/interpolate/tests/test_bsplines.py:949]: Please match the raise statements. It will help see the logic is correct.

**Review by tupui — COMMENTED** [review]
Thanks @ev-br, I have a suggestion.

File: scipy/interpolate/tests/test_bsplines.py (hunk 1 above)
Comment (ev-br) [reply scipy/interpolate/tests/test_bsplines.py]: OK, done. 

FTR, I do not believe having a `match=...`  clause is always a requirement. The readability improvement is often marginal at best, and the error messages themselves are not a part of the API. IOW, rephrasing the error message should not invalidate tests IMO.

File: scipy/interpolate/tests/test_bsplines.py [hunk 2]
```diff
@@ -942,6 +942,29 @@ def test_linear(self):
...
+        x = [0, 1, 2, 3, 4, 5]
+        y = [0, 1, 2, 3, 4, 5, 6, 7]
+        with assert_raises(ValueError, match="Shapes of x"):
+            make_interp_spline(x, y, k=k)
+
+    @pytest.mark.parametrize('k', [0, 1, 2, 3])
+    def test_broken_x(self, k):
+        x = [0, 1, 1, 2, 3, 4]      # duplicates
+        y = [0, 1, 2, 3, 4, 5]
+        with assert_raises(ValueError, match="Expect x to not have duplicates"):
```
Comment (j-bowhay) [inline scipy/interpolate/tests/test_bsplines.py:956]:
```suggestion
        with assert_raises(ValueError, match="x to not have duplicates"):
```
line too long, can just shorten to this

File: scipy/interpolate/tests/test_bsplines.py (hunk 1 above)
Comment (WarrenWeckesser) [reply scipy/interpolate/tests/test_bsplines.py]: > ... and the error messages themselves are not a part of the API.

I've asked about this before, and your statement reflects my understanding too:  the precise wording of the message in an exception is not guaranteed to remain unchanged.  For our tests, that is not a problem.  If a message is changed, and an existing test relies on the old message, we just fix the test to match the new message.

> I do not believe having a `match=...` clause is always a requirement. 

I think you are correct, we have never made this a *requirement*.  But personally, I think it is valuable enough (especially with a generic exception such as `ValueError`) that is worthwhile considering it to be a "best practice", so we don't pass a test that is actually raising an exception for an unexpected reason.

Comment (WarrenWeckesser) [reply scipy/interpolate/tests/test_bsplines.py]: Note the developer guidelines state that use `match` is the "preferred style": [Testing that expected exceptions are raised](https://docs.scipy.org/doc/scipy/dev/missing-bits.html#testing-that-expected-exceptions-are-raised)

Comment (tupui) [reply scipy/interpolate/tests/test_bsplines.py]: Sure but in this case as we have a few code paths to raise, so we want to make sure the correct path is used. E.g. ou could have another path raising and think you are raising using the last raise while you are not.

Comment (WarrenWeckesser) [reply scipy/interpolate/tests/test_bsplines.py]: @tupui, if you are responding to my comments, then we are in agreement.

Comment (ev-br) [reply scipy/interpolate/tests/test_bsplines.py]: I don't disagree in general, as long as it's indeed a *recommendation*, which PR authors and reviewers can apply or not apply depending on their judgement of a particular patch. In my experience, this does sometimes help finding edge cases, while in others it only leads to unnecessary  churn with little benefit. Thus, a recommendation not a requirement.

All that said, the comment is addressed, and after two CI flushes and a helpful suggestion from Jake (thanks!), this CI is green modulo what I believe is an unrelated issue with some doc page.

**Review by tupui — APPROVED** [review]
We are all good, thanks @ev-br et al.

**Comment by tylerjereddy** [comment]
The commit can be cleanly cherry-picked onto the maintenance branch so tentatively adding a backport label given this was considered a bug.

**Comment by ev-br** [comment]
This is a bug, yes. This one was present for several years, from pre scipy 1.1 at least. So, not new, not a regression. Whether to backport it or not, is up to you.

**Pull request merged**